File: 프로그래머스/1/172928. 공원 산책/공원 산책.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int[] solution(string[] park, string[] routes) {
        int n = park.Length;
        int m = park[0].Length;
        
        int sX = 0; int sY = 0;
        for(int i = 0; i < n; i++)
        {
            for(int j = 0; j < m; j++)
            {
                char c = park[i][j];
                if(c == 'S')
                {
                    sX = i;
                    sY = j;
                    break;
                }
            }
        }
        
        var dx = new Dictionary<string, int> { ["N"] = -1, ["S"] = 1, ["W"] = 0, ["E"] = 0 };
        var dy = new Dictionary<string, int> { ["N"] = 0, ["S"] = 0, ["W"] = -1, ["E"] = 1 }; 
        
        foreach(var route in routes)
        {
            var parts = route.Split(' ');
            string d = parts[0];
            int dist = int.Parse(parts[1]);
            
            int nx = sX;
            int ny = sY;
            bool isOK = true;
            for(int i = 0; i < dist; i++)
            {
                nx += dx[d];
                ny += dy[d];
                
                if(nx < 0 || nx >= n || ny < 0 || ny >= m || park[nx][ny] == 'X')
                {
                    isOK = false;
                    break;
                }
            }
            if(isOK) {
                sX = nx;
                sY = ny;
            }
        }
        return new int[] {sX, sY};
    }
}

File: 프로그래머스/1/172928. 공원 산책/공원 산책.cpp
#include <string>
#include <vector>

using namespace std;

int sy, sx;

vector<int> solution(vector<string> park, vector<string> routes) {
    for(int i = 0; i < park.size(); i++)
    {
        for(int j = 0; j < park[0].size(); j++)
        {
            if(park[i][j] == 'S')
            {
                sy = i;
                sx = j;
            }
        }
    }
    
    for(string route : routes)
    {
        string dir = route.substr(0, 1);
        int cnt = stoi(route.substr(2));
        
        int ny = sy;
        int nx = sx;
        
        bool check = true;
        
        for(int i = 0; i < cnt; i++)
        {
            if(dir == "N") ny--;
            else if(dir == "S") ny++;
            else if(dir == "W") nx--;
            else nx++;
            
            if(ny < 0 || nx < 0 || ny >= park.size() || nx >= park[0].size())
            {
                check = false;
                break;
            }
            if(park[ny][nx] == 'X')
            {
                check = false;
                break;
            }
        }
        if(check)
        {
            sy = ny;
            sx = nx;
        }   
    }
    return {sy, sx};
}